d3a_api_client/redis_market.py

```diff
@@ -1,23 +1,69 @@
 import logging
 import json
 import traceback
+import uuid
 from slugify import slugify
-
 from redis import StrictRedis
 from concurrent.futures.thread import ThreadPoolExecutor
 
 from d3a_interface.utils import wait_until_timeout_blocking
 from d3a_api_client.constants import MAX_WORKER_THREADS
+from d3a_api_client.utils import RedisAPIException
 
 
 class RedisMarketClient:
-    def __init__(self, area_id, redis_url='redis://localhost:6379'):
-        self.area_id = slugify(area_id, to_lower=True)
+    def __init__(self, area_id, redis_url='redis://localhost:6379', autoregister=True):
+        self.area_slug = slugify(area_id, to_lower=True)
+        self.area_uuid = None
         self.redis_db = StrictRedis.from_url(redis_url)
         self.pubsub = self.redis_db.pubsub()
         self._subscribe_to_response_channels()
         self._blocking_command_responses = {}
+        self._transaction_id_buffer = []
         self.executor = ThreadPoolExecutor(max_workers=MAX_WORKER_THREADS)
+        self.is_active = False
+        if autoregister:
+            self.register(is_blocking=True)
+
+    def register(self, is_blocking=False):
+        logging.info(f"Trying to register {self.area_slug}")
+        if self.is_active:
+            raise RedisAPIException(f'API is already registered to the market.')
+        data = {"name": self.area_slug, "transaction_id": str(uuid.uuid4())}
+        self._blocking_command_responses["register"] = data
+        self.redis_db.publish(f'{self.area_slug}/register_participant', json.dumps(data))
+
+        if is_blocking:
+            try:
+                wait_until_timeout_blocking(lambda: self.is_active, timeout=120)
+            except AssertionError:
+                raise RedisAPIException(
+                    f'API registration process timed out. Server will continue processing your '
+                    f'request on the background and will notify you as soon as the registration '
+                    f'has been completed.')
+
+    def unregister(self, is_blocking=False):
+        logging.info(f"Trying to unregister {self.area_slug}")
+
+        if not self.is_active:
+            raise RedisAPIException(f'API is already unregistered from the market.')
+
+        data = {"name": self.area_slug, "transaction_id": str(uuid.uuid4())}
+        self._blocking_command_responses["unregister"] = data
+        self.redis_db.publish(f'{self.area_slug}/unregister_participant', json.dumps(data))
+
+        if is_blocking:
+            try:
+                wait_until_timeout_blocking(lambda: not self.is_active, timeout=120)
+            except AssertionError:
+                raise RedisAPIException(
+                    f'API unregister process timed out. Server will continue processing your '
+                    f'request on the background and will notify you as soon as the unregistration '
+                    f'has been completed.')
```

Comment:
Please abstract into a method, I see code  [repetition](https://github.com/gridsingularity/d3a-api-client/pull/54/files#diff-0b61f3c71a7c84091920aa916ac35dc5R36-R43). Thanks!

Reply:
hmm, not really, the line you are pointing to is for the registration process, not for unregistering. IMO not worth the effort.

Reply:
Okay, skip it.